base/pointer.jl

```diff
@@ -98,32 +98,122 @@ unsafe_wrap(Atype::Union{Type{Array},Type{Array{T}},Type{Array{T,N}}},
 
 """
     unsafe_load(p::Ptr{T}, i::Integer=1)
+    unsafe_load(p::Ptr{T}, order::Symbol)
+    unsafe_load(p::Ptr{T}, i::Integer, order::Symbol)
 
 Load a value of type `T` from the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1]`.
+This is equivalent to the C expression `p[i-1]`. Optionally, an ordering can be provided.
 
 The `unsafe` prefix on this function indicates that no validation is performed on the
 pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program or return garbage answers. Unlike C, dereferencing
 memory region allocated as different type may be valid provided that the types are compatible.
+
+!!! compat "Julia 1.10"
+     The `order` argument is available as of Julia 1.10.
 """
 unsafe_load(p::Ptr, i::Integer=1) = pointerref(p, Int(i), 1)
+unsafe_load(p::Ptr, order::Symbol) = atomic_pointerref(p, order)
+function unsafe_load(p::Ptr, i::Integer, order::Symbol)
+    unsafe_load(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), order)
+end
 
 """
     unsafe_store!(p::Ptr{T}, x, i::Integer=1)
+    unsafe_store!(p::Ptr{T}, x, order::Symbol)
+    unsafe_store!(p::Ptr{T}, x, i::Integer, order::Symbol)
 
 Store a value of type `T` to the address of the `i`th element (1-indexed) starting at `p`.
-This is equivalent to the C expression `p[i-1] = x`.
+This is equivalent to the C expression `p[i-1] = x`. Optionally, an ordering can be
+provided.
 
 The `unsafe` prefix on this function indicates that no validation is performed on the
 pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
 that referenced memory is not freed or garbage collected while invoking this function.
 Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
 different type may be valid provided that that the types are compatible.
+
+!!! compat "Julia 1.10"
+     The `order` argument is available as of Julia 1.10.
 """
 unsafe_store!(p::Ptr{Any}, @nospecialize(x), i::Integer=1) = pointerset(p, x, Int(i), 1)
 unsafe_store!(p::Ptr{T}, x, i::Integer=1) where {T} = pointerset(p, convert(T,x), Int(i), 1)
+unsafe_store!(p::Ptr, x, order::Symbol) = atomic_pointerset(p, x, order)
+function unsafe_store!(p::Ptr, x, i::Integer, order::Symbol)
+    unsafe_store!(p + (aligned_sizeof(eltype(p)) * (Int(i) - 1)), x, order)
+end
+
+"""
+    unsafe_modify!(p::Ptr{T}, op, x, [order::Symbol]) -> Pair
+
+These atomically perform the operations to get and set a field after applying
+the function `op`.
+
+    y = unsafe_load(p)
+    z = op(y, x)
+    unsafe_store!(p, z)
+    return y => z
+
```

Comment:
nit: add the standard boiler plate warning:
```suggestion

The `unsafe` prefix on this function indicates that no validation is performed on the
pointer `p` to ensure that it is valid. Like C, the programmer is responsible for ensuring
that referenced memory is not freed or garbage collected while invoking this function.
Incorrect usage may segfault your program. Unlike C, storing memory region allocated as
different type may be valid provided that that the types are compatible.

```